Hard-coded proxy connect timeout in REST client overrides connect timeout of REST client.

### Describe the bug

Due to a bad actor in an external network, my team and I have discovered that vert.x does not enable configuration of the the connect timeout for proxy clients. See issue https://github.com/eclipse-vertx/vert.x/issues/5634

The Quarkus documentation outlines a connect timeout of 15 seconds, but in practice, the HTTP proxy client enforces a limit of 10 seconds, which cannot be configured or changed.

### Expected behavior

If a HTTP proxy (or proxied server) requires 11 seconds to connect, and the Quarkus REST Client has a connect timeout of 15 seconds, I would expect the connection to be established.

### Actual behavior

The HTTP Proxy client prematurely terminates the connection after 10 seconds.

```
jakarta.ws.rs.ProcessingException: io.netty.handler.proxy.ProxyConnectException: http, basic, /[<IPv6 address>]:7071 => <the bad actor>, timeout
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$2.accept(ClientSendRequestHandler.java:225)
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$2.accept(ClientSendRequestHandler.java:217)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at io.smallrye.mutiny.helpers.UniCallbackSubscriber.onFailure(UniCallbackSubscriber.java:62)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:55)
	at org.jboss.resteasy.reactive.client.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:37)
	at io.vertx.core.impl.future.FutureImpl$4.onFailure(FutureImpl.java:188)
	at io.vertx.core.impl.future.FutureBase.lambda$emitFailure$1(FutureBase.java:75)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:312)
	at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:168)
	at io.vertx.core.impl.future.FutureBase.emitFailure(FutureBase.java:72)
	at io.vertx.core.impl.future.FutureImpl.tryFail(FutureImpl.java:278)
	at io.vertx.core.http.impl.HttpClientImpl.lambda$doRequest$4(HttpClientImpl.java:398)
	at io.vertx.core.net.impl.pool.Endpoint.lambda$getConnection$0(Endpoint.java:52)
	at io.vertx.core.http.impl.SharedClientHttpStreamEndpoint$Request.handle(SharedClientHttpStreamEndpoint.java:162)
	at io.vertx.core.http.impl.SharedClientHttpStreamEndpoint$Request.handle(SharedClientHttpStreamEndpoint.java:123)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:342)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:335)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool$ConnectFailed$1.run(SimpleConnectionPool.java:380)
	at io.vertx.core.net.impl.pool.Task.runNextTasks(Task.java:43)
	at io.vertx.core.net.impl.pool.CombinerExecutor.submit(CombinerExecutor.java:91)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.execute(SimpleConnectionPool.java:244)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.lambda$connect$2(SimpleConnectionPool.java:258)
	at io.vertx.core.http.impl.SharedClientHttpStreamEndpoint.lambda$connect$2(SharedClientHttpStreamEndpoint.java:104)
	at io.vertx.core.impl.future.FutureImpl$4.onFailure(FutureImpl.java:188)
	at io.vertx.core.impl.future.FutureBase.emitFailure(FutureBase.java:81)
	at io.vertx.core.impl.future.FutureImpl.tryFail(FutureImpl.java:278)
	at io.vertx.core.impl.future.Composition$1.onFailure(Composition.java:66)
	at io.vertx.core.impl.future.FutureBase.emitFailure(FutureBase.java:81)
	at io.vertx.core.impl.future.FailedFuture.addListener(FailedFuture.java:98)
	at io.vertx.core.impl.future.Composition.onFailure(Composition.java:55)
	at io.vertx.core.impl.future.FutureBase.emitFailure(FutureBase.java:81)
	at io.vertx.core.impl.future.FutureImpl.tryFail(FutureImpl.java:278)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:342)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:335)
	at io.vertx.core.net.impl.NetClientImpl.failed(NetClientImpl.java:351)
	at io.vertx.core.net.impl.NetClientImpl.lambda$connectInternal2$6(NetClientImpl.java:323)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:603)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:570)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:505)
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:649)
	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:642)
	at io.netty.util.concurrent.DefaultPromise.setFailure(DefaultPromise.java:123)
	at io.vertx.core.net.impl.ChannelProvider$3$1.exceptionCaught(ChannelProvider.java:235)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:317)
	at io.netty.handler.proxy.ProxyHandler.failPendingWritesAndClose(ProxyHandler.java:356)
	at io.netty.handler.proxy.ProxyHandler.setConnectFailure(ProxyHandler.java:349)
	at io.netty.handler.proxy.ProxyHandler.access$100(ProxyHandler.java:41)
	at io.netty.handler.proxy.ProxyHandler$2.run(ProxyHandler.java:201)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:156)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.netty.handler.proxy.ProxyConnectException: http, basic, /[<IPv6 address>]:7071 => <the bad actor>, timeout
	... 11 more
```

### How to Reproduce?

Use a HTTP proxy server and establish a connection, where one of the involved parties takes over 10 seconds to establish the connection.

(Or see https://github.com/eclipse-vertx/vert.x/issues/5634#issuecomment-3079058609 )

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Vertx have already begun work on a fix, enabling a new `connectTimeout` configuration for the proxy client. See PR https://github.com/eclipse-vertx/vert.x/pull/5650

The developers have stated interest in backporting the fix. https://github.com/eclipse-vertx/vert.x/issues/5634#issuecomment-3084176444

--- a/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java
+++ b/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java
@@ -339,7 +339,7 @@ public interface RestClientsConfig {
          * <li>{@code none} - no additional logging</li>
          * </ul>
          *
-         * This property is applicable to reactive REST clients only.
+         * This property is not applicable to the Quarkus RESTEasy client (provided by the quarkus-resteasy-client dependency).
          */
         Optional<String> scope();
 
@@ -348,7 +348,7 @@ public interface RestClientsConfig {
          * <p>
          * By default, set to 100.
          * <p>
-         * This property is applicable to reactive REST clients only.
+         * This property is not applicable to the Quarkus RESTEasy client (provided by the quarkus-resteasy-client dependency).
          */
         @WithDefault("100")
         Integer bodyLimit();
@@ -358,7 +358,7 @@ public interface RestClientsConfig {
         /**
          * The max HTTP chunk size (8096 bytes by default).
          * <p>
-         * This property is applicable to reactive REST clients only.
+         * This property is not applicable to the Quarkus RESTEasy client (provided by the quarkus-resteasy-client dependency).
          *
          * @deprecated Use {@code quarkus.rest-client.max-chunk-size} instead
          */
@@ -417,8 +417,7 @@ public interface RestClientsConfig {
          * url.
          * The override is done using the REST Client class name configuration syntax.
          * <p>
-         * This property is not applicable to the RESTEasy Client, only the Quarkus Rest client (formerly RESTEasy Reactive
-         * client).
+         * This property is not applicable to the Quarkus RESTEasy client (provided by the quarkus-resteasy-client dependency).
          */
         Optional<String> overrideUri();
 

--- a/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java
+++ b/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java
@@ -1,6 +1,7 @@
 package io.quarkus.restclient.config;
 
 import java.io.InputStream;
+import java.time.Duration;
 import java.util.Map;
 import java.util.Optional;
 import java.util.OptionalInt;
@@ -88,6 +89,16 @@ public interface RestClientsConfig {
      * This property is not applicable to the RESTEasy Client.
      */
     Optional<String> nonProxyHosts();
+
+    /**
+     * Proxy connection timeout
+     * <p>
+     * Can be overwritten by client-specific settings.
+     * <p>
+     * This property is not applicable to the RESTEasy Client.
+     */
+    @ConfigDocDefault("10s")
+    Optional<Duration> proxyConnectTimeout();
 
     /**
      * A timeout in milliseconds that REST clients should wait to connect to the remote endpoint.
@@ -483,6 +494,14 @@ public interface RestClientsConfig {
          * This property is not applicable to the RESTEasy Client.
          */
         Optional<String> nonProxyHosts();
+
+        /**
+         * Proxy connection timeout
+         * <p>
+         * This property is not applicable to the RESTEasy Client.
+         */
+        @ConfigDocDefault("10s")
+        Optional<Duration> proxyConnectTimeout();
 
         /**
          * An enumerated type string value with possible values of "MULTI_PAIRS" (default), "COMMA_SEPARATED",

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/QuarkusRestClientBuilder.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/QuarkusRestClientBuilder.java
@@ -3,6 +3,7 @@ package io.quarkus.rest.client.reactive;
 import java.net.URI;
 import java.net.URL;
 import java.security.KeyStore;
+import java.time.Duration;
 import java.util.ServiceLoader;
 import java.util.concurrent.TimeUnit;
 
@@ -225,6 +226,14 @@ public interface QuarkusRestClientBuilder extends Configurable<QuarkusRestClient
      * @return the current builder
      */
     QuarkusRestClientBuilder nonProxyHosts(String nonProxyHosts);
+
+    /**
+     * Specifies the connect timeout for the proxy connection
+     *
+     * @param connectTimeout proxy connect timeout.
+     * @return the current builder
+     */
+    QuarkusRestClientBuilder proxyConnectTimeout(Duration connectTimeout);
 
     /**
      * Specifies the URI formatting style to use when multiple query parameter values are passed to the client.

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/QuarkusRestClientBuilderImpl.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/QuarkusRestClientBuilderImpl.java
@@ -3,6 +3,7 @@ package io.quarkus.rest.client.reactive.runtime;
 import java.net.URI;
 import java.net.URL;
 import java.security.KeyStore;
+import java.time.Duration;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
@@ -124,6 +125,12 @@ public class QuarkusRestClientBuilderImpl implements QuarkusRestClientBuilder {
     @Override
     public QuarkusRestClientBuilder nonProxyHosts(String nonProxyHosts) {
         delegate.nonProxyHosts(nonProxyHosts);
+        return this;
+    }
+
+    @Override
+    public QuarkusRestClientBuilder proxyConnectTimeout(Duration connectTimeout) {
+        delegate.proxyConnectTimeout(connectTimeout);
         return this;
     }
 

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientBuilderImpl.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientBuilderImpl.java
@@ -7,6 +7,7 @@ import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.security.KeyStore;
+import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
@@ -81,6 +82,7 @@ public class RestClientBuilderImpl implements RestClientBuilder {
     private String proxyUser;
     private String proxyPassword;
     private String nonProxyHosts;
+    private Duration proxyConnectTimeout;
 
     private ClientLogger clientLogger;
     private LoggingScope loggingScope;
@@ -234,6 +236,11 @@ public class RestClientBuilderImpl implements RestClientBuilder {
 
     public RestClientBuilderImpl nonProxyHosts(String nonProxyHosts) {
         this.nonProxyHosts = nonProxyHosts;
+        return this;
+    }
+
+    public RestClientBuilderImpl proxyConnectTimeout(Duration proxyConnectTimeout) {
+        this.proxyConnectTimeout = proxyConnectTimeout;
         return this;
     }
 
@@ -568,11 +575,16 @@ public class RestClientBuilderImpl implements RestClientBuilder {
         }
 
         if (proxyHost != null) {
-            configureProxy(proxyHost, proxyPort, proxyUser, proxyPassword, nonProxyHosts);
+            configureProxy(proxyHost, proxyPort, proxyUser, proxyPassword, nonProxyHosts, proxyConnectTimeout);
         } else if (restClients.proxyAddress().isPresent()) {
             HostAndPort globalProxy = ProxyAddressUtil.parseAddress(restClients.proxyAddress().get());
-            configureProxy(globalProxy.host, globalProxy.port, restClients.proxyUser().orElse(null),
-                    restClients.proxyPassword().orElse(null), restClients.nonProxyHosts().orElse(null));
+            configureProxy(
+                    globalProxy.host,
+                    globalProxy.port,
+                    restClients.proxyUser().orElse(null),
+                    restClients.proxyPassword().orElse(null),
+                    restClients.nonProxyHosts().orElse(null),
+                    restClients.proxyConnectTimeout().orElse(null));
         }
 
         if (!clientBuilder.getConfiguration().hasProperty(QuarkusRestClientProperties.MULTIPART_ENCODER_MODE)) {
@@ -604,7 +616,7 @@ public class RestClientBuilderImpl implements RestClientBuilder {
     }
 
     private void configureProxy(String proxyHost, Integer proxyPort, String proxyUser, String proxyPassword,
-            String nonProxyHosts) {
+            String nonProxyHosts, Duration proxyConnectTimeout) {
         if (proxyHost != null) {
             clientBuilder.proxy(proxyHost, proxyPort);
             if (proxyUser != null && proxyPassword != null) {
@@ -614,6 +626,10 @@ public class RestClientBuilderImpl implements RestClientBuilder {
 
             if (nonProxyHosts != null) {
                 clientBuilder.nonProxyHosts(nonProxyHosts);
+            }
+
+            if (proxyConnectTimeout != null) {
+                clientBuilder.proxyConnectTimeout(proxyConnectTimeout);
             }
         }
     }

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientCDIDelegateBuilder.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientCDIDelegateBuilder.java
@@ -194,6 +194,8 @@ public class RestClientCDIDelegateBuilder<T> {
             oneOf(restClientConfig.proxyUser(), configRoot.proxyUser()).ifPresent(builder::proxyUser);
             oneOf(restClientConfig.proxyPassword(), configRoot.proxyPassword()).ifPresent(builder::proxyPassword);
             oneOf(restClientConfig.nonProxyHosts(), configRoot.nonProxyHosts()).ifPresent(builder::nonProxyHosts);
+            oneOf(restClientConfig.proxyConnectTimeout(), configRoot.proxyConnectTimeout())
+                    .ifPresent(builder::proxyConnectTimeout);
         }
     }
 

--- a/extensions/resteasy-reactive/rest-client/runtime/src/test/java/io/quarkus/rest/client/reactive/runtime/RestClientCDIDelegateBuilderTest.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/test/java/io/quarkus/rest/client/reactive/runtime/RestClientCDIDelegateBuilderTest.java
@@ -25,6 +25,7 @@ import java.security.KeyStore;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
 import java.security.cert.CertificateException;
+import java.time.Duration;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -134,6 +135,7 @@ class RestClientCDIDelegateBuilderTest {
         verify(restClientBuilderMock).proxyUser("proxyUser1");
         verify(restClientBuilderMock).proxyPassword("proxyPassword1");
         verify(restClientBuilderMock).nonProxyHosts("nonProxyHosts1");
+        verify(restClientBuilderMock).proxyConnectTimeout(Duration.ofSeconds(20));
         verify(restClientBuilderMock).connectTimeout(100, MILLISECONDS);
         verify(restClientBuilderMock).readTimeout(101, MILLISECONDS);
         verify(restClientBuilderMock).userAgent("agent1");
@@ -188,6 +190,7 @@ class RestClientCDIDelegateBuilderTest {
         verify(restClientBuilderMock).proxyUser("proxyUser2");
         verify(restClientBuilderMock).proxyPassword("proxyPassword2");
         verify(restClientBuilderMock).nonProxyHosts("nonProxyHosts2");
+        verify(restClientBuilderMock).proxyConnectTimeout(Duration.ofSeconds(30));
         verify(restClientBuilderMock).connectTimeout(200, MILLISECONDS);
         verify(restClientBuilderMock).readTimeout(201, MILLISECONDS);
         verify(restClientBuilderMock).userAgent("agent2");
@@ -215,6 +218,7 @@ class RestClientCDIDelegateBuilderTest {
         rootConfig.put("quarkus.rest-client.proxy-user", "proxyUser2");
         rootConfig.put("quarkus.rest-client.proxy-password", "proxyPassword2");
         rootConfig.put("quarkus.rest-client.non-proxy-hosts", "nonProxyHosts2");
+        rootConfig.put("quarkus.rest-client.proxy-connect-timeout", "30s");
         rootConfig.put("quarkus.rest-client.connect-timeout", "200");
         rootConfig.put("quarkus.rest-client.read-timeout", "201");
         rootConfig.put("quarkus.rest-client.user-agent", "agent2");
@@ -250,6 +254,7 @@ class RestClientCDIDelegateBuilderTest {
         clientConfig.put("quarkus.rest-client." + restClientName + ".proxy-user", "proxyUser1");
         clientConfig.put("quarkus.rest-client." + restClientName + ".proxy-password", "proxyPassword1");
         clientConfig.put("quarkus.rest-client." + restClientName + ".non-proxy-hosts", "nonProxyHosts1");
+        clientConfig.put("quarkus.rest-client." + restClientName + ".proxy-connect-timeout", "20s");
         clientConfig.put("quarkus.rest-client." + restClientName + ".connect-timeout", "100");
         clientConfig.put("quarkus.rest-client." + restClientName + ".read-timeout", "101");
         clientConfig.put("quarkus.rest-client." + restClientName + ".user-agent", "agent1");

--- a/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java
+++ b/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java
@@ -9,6 +9,7 @@ import java.security.KeyStore;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
 import java.security.cert.CertificateException;
+import java.time.Duration;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -56,6 +57,7 @@ public class ClientBuilderImpl extends ClientBuilder {
     private String proxyPassword;
     private String proxyUser;
     private String nonProxyHosts;
+    private Duration proxyConnectTimeout;
 
     private boolean followRedirects;
 
@@ -254,6 +256,9 @@ public class ClientBuilderImpl extends ClientBuilder {
                 }
                 if (proxyUser != null && !proxyUser.isBlank()) {
                     proxyOptions.setUsername(proxyUser);
+                }
+                if (proxyConnectTimeout != null) {
+                    proxyOptions.setConnectTimeout(proxyConnectTimeout);
                 }
                 options.setProxyOptions(proxyOptions);
                 configureNonProxyHosts(options, nonProxyHosts);
@@ -489,6 +494,11 @@ public class ClientBuilderImpl extends ClientBuilder {
 
     public ClientBuilderImpl nonProxyHosts(String nonProxyHosts) {
         this.nonProxyHosts = nonProxyHosts;
+        return this;
+    }
+
+    public ClientBuilderImpl proxyConnectTimeout(Duration proxyConnectTimeout) {
+        this.proxyConnectTimeout = proxyConnectTimeout;
         return this;
     }
 }

